### PR TITLE
lomiri.morph-browser: init at 1.1.0

### PR DESCRIPTION
**Pull request by OPNA2608**
## Description of changes

Working towards #99090.

[Morph Browser](https://gitlab.com/ubports/development/core/morph-browser) is the Lomiri web browser. Last time I checked, neither Firefox nor Chrome work under Lomiri (Wayland), so this might be necessary for everyday use.

image

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by OPNA2608** [comment]
Works as-is, but I'll submit patches to upstream to improve some things for our needs.

**Comment by OPNA2608** [comment]
Mostly done, but:
- Patching-out always-enabled tests could be submitted upstream, but submissions relating to opting out of tests are stalled until upstream [decides on how to approach it w/r/t naming the CMake option](https://gitlab.com/ubports/development/core/history-service/-/merge_requests/39#note_1707114350).
- I want to test the fix to `X-Lomiri-Splash-Image` first before undrafting (only speculative, based on [staring](https://gitlab.com/ubports/development/core/lomiri-app-launch/-/blob/1a7a0ef6a8ab966b664f3515aba539b422f35849/liblomiri-app-launch/application-impl-legacy.cpp) at [code](https://gitlab.com/ubports/development/core/lomiri-app-launch/-/blob/1a7a0ef6a8ab966b664f3515aba539b422f35849/liblomiri-app-launch/application-info-desktop.cpp#L421)).

**Comment by OPNA2608** [comment]
For getting proper graphics in testing (normally Lomiri / its NixOS module would set this up):

```sh
env UITK_ICON_THEME=suru XDG_DATA_DIRS="$(nix-build --no-out-link . -A lomiri.suru-icon-theme)"/share:$XDG_DATA_DIRS "$(nix-build --no-out-link . -A lomiri.morph-browser)"/bin/morph-browser
```

I also have a VM subtest that I wrote for it as part of my Lomiri shell VM test. Not sure if it would make sense to pull that out & into a separate Morph Browser test on another DE, or if doing that as part of one big Lomiri test & adding that to Morph's `passthru.tests` would be good enough.
```py
    with subtest("morph browser works"):
        machine.send_key("meta_l-a") # app drawer
        machine.wait_for_text("Morph") # morph-browser should be listed
        machine.send_chars("Morph\n") # launch it
        machine.wait_for_text("Bookmarks") # new page has bookmarks section
        maximise_new_window(machine)
        machine.sleep(1) # VM is sluggish
        machine.screenshot("morph_open")

        # Try to display Valgrind documentation
        machine.send_chars("file://${pkgs.valgrind.doc}/share/doc/valgrind/html/index.html\n")
        machine.wait_for_text("Valgrind")
        machine.screenshot("morph_htmlcontent")

        machine.send_key("alt-f4")
```

File: pkgs/desktops/lomiri/applications/morph-browser/default.nix [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    lomiri-action-api
+    lomiri-ui-extras
+    lomiri-ui-toolkit
+    qtquickcontrols2
+    qtsystems
+  ];
+
+  nativeCheckInputs = [
+    (python3.withPackages (ps: with ps; [
+      flake8
```
Comment (SuperSandro2000) [inline pkgs/desktops/lomiri/applications/morph-browser/default.nix:93]: We don't want to run linters as part of the tests

**Review by SuperSandro2000 — DISMISSED** [review]

**Comment by OPNA2608** [comment]
Ping @SuperSandro2000, anything else?

**Comment by pbsds** [comment]
I'm getting a timeout on the nixosTest when merged onto master

**Comment by OPNA2608** [comment]
Seems to be an issue with the latest imagemagick bump (https://github.com/NixOS/nixpkgs/pull/288208), bisected within the `nixpkgs-review` shell.
```
git bisect start
# Status: warte auf guten und schlechten Commit
# bad: [6c8d5bb7169c8bad0a5c203279524df1e7c673e6] morph-browser
git bisect bad 6c8d5bb7169c8bad0a5c203279524df1e7c673e6
# Status: warte auf gute(n) Commit(s), schlechter Commit bekannt
# good: [fbe2d5c2852179db37f7a6af92052309fad7b5ea] Merge pull request #285120 from OPNA2608/fix/lomiri-ui-toolkit-versioned-QML_IMPORTS_PATH
git bisect good fbe2d5c2852179db37f7a6af92052309fad7b5ea
# good: [e9b728e4fdbd2efee1eebad2e6e591b253e7b2f8] Merge pull request #284580 from D3vil0p3r/patch-5
git bisect good e9b728e4fdbd2efee1eebad2e6e591b253e7b2f8
# bad: [60abd7c1f985d0679675ec860024e7fa16f229ee] Merge pull request #288138 from ambroisie/tandoor-recipes-manage-script
git bisect bad 60abd7c1f985d0679675ec860024e7fa16f229ee
# bad: [db19b2874696e9d0adbd0078735831eab7b51c01] Merge pull request #285249 from GaetanLepage/torch
git bisect bad db19b2874696e9d0adbd0078735831eab7b51c01
# good: [b8a84834b23cc1edd649e9921bf7d5f5d9b4118d] Merge pull request #287766 from bobby285271/upd/gnome
git bisect good b8a84834b23cc1edd649e9921bf7d5f5d9b4118d
# good: [f7152e2e18a80a61e507c4be485ca1289dc8cffd] Merge pull request #287924 from wegank/msolve-bump
git bisect good f7152e2e18a80a61e507c4be485ca1289dc8cffd
# good: [113bb026d5d3d873a2c52787e26af10ebb5f30b0] Merge pull request #286850 from khaneliman/sketchybar
git bisect good 113bb026d5d3d873a2c52787e26af10ebb5f30b0
# bad: [88e66deef19b2b920a4bae96a9a7cb5f4eafb9a9] Merge pull request #287572 from trofi/invoiceplane-fix-hash
git bisect bad 88e66deef19b2b920a4bae96a9a7cb5f4eafb9a9
# bad: [a9ec8d4aa34ad005597cf8e67018c1112c0926e7] Merge pull request #287862 from D3vil0p3r/patch-12
git bisect bad a9ec8d4aa34ad005597cf8e67018c1112c0926e7
# good: [b54cebbc14d59a8768a878ccd0e0f0deaff099da] Merge pull request #286580 from SuperSamus/wine-update
git bisect good b54cebbc14d59a8768a878ccd0e0f0deaff099da
# bad: [36ab9337c0f92b09ce06146b2fd4f3b061780b77] Merge pull request #288208 from r-ryantm/auto-update/imagemagick
git bisect bad 36ab9337c0f92b09ce06146b2fd4f3b061780b77
# good: [2e26817d1275a81fed48318e4b26eb0739f0340c] Merge pull request #284707 from ocfox/tide
git bisect good 2e26817d1275a81fed48318e4b26eb0739f0340c
# good: [e03ad4b2f916504bcedeff945486d119ff71c2ea] Merge pull request #288213 from dotlambda/stem-1.8.3
git bisect good e03ad4b2f916504bcedeff945486d119ff71c2ea
# bad: [6a79a442b49967fd8ccae8eeb9f03a066d74eac4] imagemagick: 7.1.1-27 -> 7.1.1-28
git bisect bad 6a79a442b49967fd8ccae8eeb9f03a066d74eac4
# first bad commit: [6a79a442b49967fd8ccae8eeb9f03a066d74eac4] imagemagick: 7.1.1-27 -> 7.1.1-28
```

I can confirm via reverting some commits that https://github.com/NixOS/nixpkgs/pull/289264 had to disable some of its OCR tests due to degraded text recognition with this new version, and that their old tests work fine on the previous version. I'll also adjust the OCR part of the tests here.

(Not sure if this is worth escalating to an issue, prolly just a part of dealing with the OCR's moodiness. Just wanted to mention the path I took for figuring this out.)

**Comment by OPNA2608** [comment]
Times out on ARM, argh.

```
(finished: subtest: morph browser launches, in 33.29 seconds)
[...]
(finished: subtest: morph browser displays HTML, in 47.96 seconds)
[...]
subtest: morph browser localisation works
machine: waiting for Neuer Tab to appear on screen
[...]
error: building of '/nix/store/1fr4cryykf8wyrh3zbvl3zxfjf1cwl5x-vm-test-run-morph-browser-standalone.drv^out' from .drv file timed out after 3600 seconds
```

Will try to find some clearer text to OCR for…

**Comment by OPNA2608** [comment]
I added various UI strings to look for, I hope it's more reliable now.

**Review by pbsds — APPROVED** [review]
Runs fine, and the derivation and desktop files LGTM

**Pull request merged**